### PR TITLE
feat: run shared tests

### PR DESCRIPTION
**Pull request by jannyHou**
### Description

connect to https://github.com/strongloop/loopback-connector-mysql/issues/386

Run shared tests from juggler v3 and v4.

#### Related issues
https://github.com/strongloop/loopback-connector-mysql/issues/386

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,1902 @@
+{
```
Comment (rmg) [inline package-lock.json]: I strongly recommend _not_ committing a `package-lock.json` to this repo. It will lock the dependencies used by CI and loopback-connector-mysql devs while _users_ of loopback-connector-mysql will get newer/updated versions of the dependencies.

This means that CI will not be testing what users are using and may missing important bugs introduced by upstream dependencies.

Comment (jannyHou) [reply package-lock.json]: True, deleted 🙂 it's auto generated by my PR, now tests run successfully, thanks!

File: .travis.yml [hunk 2]
```diff
@@ -1,6 +1,5 @@
 language: node_js
 node_js:
-  - "6"
   - "8"
   - "10"
```
Comment (bajtos) [inline .travis.yml]: Can you add Node.js 12 while changing `.travis.yml`? Not a big deal, we can add it later in a new PR.

File: lib/mysql.js [hunk 3]
```diff
@@ -444,6 +444,12 @@ MySQL.prototype.fromColumnValue = function(prop, val) {
           lat: val.y,
         };
         break;
+      case 'ObjectID':
```
Comment (bajtos) [inline lib/mysql.js]: Interesting. Why do we need to deal with `ObjectID`, considering that it's a type specific to MongoDB? Is this related to changes made in juggler v4?

File: lib/mysql.js [hunk 4]
```diff
@@ -453,10 +459,6 @@ MySQL.prototype.fromColumnValue = function(prop, val) {
         }
         break;
       default:
-        if (!Array.isArray(prop.type) && !prop.type.modelName) {
-          // Do not convert array and model types
```
Comment (bajtos) [inline lib/mysql.js]: If you remove this `default` case, does it mean that we are no longer going to convert other types? What are the implications for existing applications, can we break any edge cases?

File: test/init.js [hunk 5]
```diff
@@ -6,8 +6,8 @@
 'use strict';
 
 module.exports = require('should');
-
-var DataSource = require('loopback-datasource-juggler').DataSource;
+const juggler = require('loopback-datasource-juggler');
+let DataSource = require('loopback-datasource-juggler').DataSource;
```
Comment (bajtos) [inline test/init.js]:
```suggestion
let DataSource = juggler.DataSource;
```

File: test/datatypes.test.js [hunk 6]
```diff
@@ -206,7 +207,8 @@ describe('MySQL specific datatypes', function() {
     var xcor, ycor;
     City.create(city1, function(err, res) {
       if (err) return done(err);
-      res.loc.should.deepEqual(city1.loc);
+      const loc_in_geo_type = new GeoPoint(city1.loc);
+      res.loc.should.deepEqual(loc_in_geo_type);
```
Comment (bajtos) [inline test/datatypes.test.js]: This change seems unrelated to the introduction of shared test suite from juggler v4. Is it needed to fix failing tests? Can we make it in a standalone PR please?

File: test/migration.test.js [hunk 7]
```diff
@@ -639,7 +639,9 @@ var getFields = function(model, cb) {
       res.forEach(function(field) {
         fields[field.Field] = field;
       });
-      cb(err, fields);
+      // The returned data are in arrays of type `RowDataPacket`,
+      // which are not objects.
+      cb(err, JSON.parse(JSON.stringify(fields)));
```
Comment (bajtos) [inline test/migration.test.js]: This change seems unrelated to the introduction of shared test suite from juggler v4. Is it needed to fix failing tests? Can we make it in a standalone PR please?

File: test/mysql.test.js [hunk 8]
```diff
@@ -88,7 +89,7 @@ describe('mysql', function() {
 
         p.content.should.be.equal(post.content);
         p.title.should.be.equal('a');
-        p.comments.should.eql(['1', '2']);
+        p.comments.should.eql(new List(['1', '2']));
```
Comment (bajtos) [inline test/mysql.test.js]: This change seems unrelated to the introduction of shared test suite from juggler v4. Is it needed to fix failing tests? Can we make it in a standalone PR please?

**Comment by jannyHou** [comment]
https://github.com/strongloop/loopback-connector-mysql/pull/390#discussion_r301130833
https://github.com/strongloop/loopback-connector-mysql/pull/390#discussion_r301130783
https://github.com/strongloop/loopback-connector-mysql/pull/390#discussion_r301130754

I submitted another PR to verify are they working with juggler v3 in

File: .travis.yml (hunk 2 above)
Comment (jannyHou) [reply .travis.yml]: @bajtos Yep I can submit another PR to add Node.js 12, I moved 6 in this PR since CI fails.

Running juggler v3 and v4 causes 20+ failures, I am not sure are they all broken by the upgrading of juggler, but this PR and https://github.com/strongloop/loopback-connector-mysql/pull/392 needs to fix them all, which is already a big amount of work, so I would like to add Node 12 in a separate PR :)

File: lib/mysql.js (hunk 3 above)
Comment (jannyHou) [reply lib/mysql.js]: @bajtos See tests in https://github.com/strongloop/loopback-connector-mysql/blob/1849507e065985ea0242440979a073bcebe4682e/test/mysql.test.js#L14, I guess it's for consistent behavior across all connectors.

File: lib/mysql.js (hunk 4 above)
Comment (jannyHou) [reply lib/mysql.js]: @bajtos The fix here is tricky, I have spent lots of time to figure out what is the expected behaviour across all the connectors:

When running tests in juggler v4, the type conversion here causes https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/datatype.test.js#L193-L199 fail. This test expects the connector returns the value from db AS IT IS.

And if you look at the postgresql connector https://github.com/strongloop/loopback-connector-postgresql/blob/master/lib/postgresql.js#L246, it also returns the value WITHOUT conversion.

Therefore I believe at least for juggler v4, the expected behaviour is connectors don't do type conversions for unknown types.

**Comment by jannyHou** [comment]
@bajtos [Those 3 tests](https://github.com/strongloop/loopback-connector-mysql/pull/390#issuecomment-509319325) are related to the v3->v4 upgrade of juggler, see PR https://github.com/strongloop/loopback-connector-mysql/pull/392, so those fixes should still be kept in this PR.

I think v4 has a breaking change for the type coercion.

**Comment by jannyHou** [comment]
@bajtos Thank you for the review, replied.

I am ok to release a new major version for upgrading to juggler v4. For the failures on the current master,  I prefer to leave it out of the scope of issue https://github.com/strongloop/loopback-connector-mysql/issues/386, we can either reopen https://github.com/strongloop/loopback-connector-mysql/issues/378 or create a new task for it :)

File: .travis.yml (hunk 2 above)
Comment (bajtos) [reply .travis.yml]: Makes sense :+1:

File: lib/mysql.js (hunk 3 above)
Comment (bajtos) [reply lib/mysql.js]: I see. `test/mysql.test.js` is not part of the shared test suite imported from juggler, so why are we fixing it in this pull request? Can you please open a new pull requests to fix problems that are unrelated to the shared test suite first?

File: lib/mysql.js (hunk 4 above)
Comment (bajtos) [reply lib/mysql.js]: I see, thank you for explaining the details to me.

**Comment by bajtos** [comment]
I am not sure if we need a new major version, leave that up to you to decide.

I think I am starting to understand what's the problem here. We are not only adding the shared test suite from juggler v4, but also upgrading the mysql-connector-specific test suite to execute against juggler v4. On `master`, these connector-specific tests are executed against juggler v3.

In that case it makes sense to keep the changes in `test/*` files as part of this pull request 👍 
- https://github.com/strongloop/loopback-connector-mysql/pull/390#discussion_r301130833
- https://github.com/strongloop/loopback-connector-mysql/pull/390#discussion_r301130783 
- https://github.com/strongloop/loopback-connector-mysql/pull/390#discussion_r301130754

File: lib/mysql.js (hunk 3 above)
Comment (jannyHou) [reply lib/mysql.js]: @bajtos Interesting 🤔 I thought you have noticed that after configuring the shared tests, the connector tests run with juggler v4(which this PR does) instead of juggler v3(the current master), and that's how all these fixes happen. 
And I thought "running connector tests with juggler v4" is what we intend to do...

**Comment by jannyHou** [comment]
@bajtos 
> I think I am starting to understand what's the problem here. We are not only adding the shared test suite from juggler v4, but also upgrading the mysql-connector-specific test suite to execute against juggler v4. On master, these connector-specific tests are executed against juggler v3.

Ah yes! Then never mind about my comment https://github.com/strongloop/loopback-connector-mysql/pull/390#discussion_r301623757.
